### PR TITLE
Retry on coroutines

### PR DESCRIPTION
**Pull request by bersace**
Hi,

I'm a happy user of retrying, and i would like to use its features to retry coroutines. I wrote a PoC at https://github.com/novafloss/jenkins-epo/blob/master/jenkins_epo/utils.py#L34 , but i would like to contribute a more sane and integrated support.

So here is my contribution :) It comes with one unit test which i hope is enough for code coverage.

What do you think of this ? I would like to make it as consistent as possible with existing code and minimal.

Regards,
Étienne

### CODE REVIEW

File: retrying.py [hunk 1]
```diff
@@ -37,16 +42,7 @@ def retry(*dargs, **dkw):
...
-
-            @six.wraps(f)
-            def wrapped_f(*args, **kw):
-                return Retrying().call(f, *args, **kw)
-
-            return wrapped_f
-
-        return wrap_simple(dargs[0])
-
+        return retry()(dargs[0])
```
Comment (bersace) [inline retrying.py:48]: Here is just a bit of code golf. I though i had to simplify this to manage `iscoroutinefunction` on decorated functions, but finally that seems unnecessary. I can let the decorator untouched if you want :)

File: retrying.py [hunk 2]
```diff
@@ -214,7 +210,84 @@ def should_reject(self, attempt):
 
         return reject
 
+    def retry_main_loop(self):
```
Comment (bersace) [inline retrying.py:222]: This generator yield each time it requires an attempt. This let `call_sync` or `call_async` call `fn` approrietly, without worrying the retry marshalling.

**Comment by bersace** [comment]
```
  File "/home/travis/build/rholder/retrying/retrying.py", line 272
    try_result = yield from fn(*args, **kwargs)
                          ^
SyntaxError: invalid syntax
```

Too bad this code, never executed on py2 and py3.3 triggers a `SyntaxError`. Do you have a pro tip on this ? Must I use `for` ? Separate file ? (i like the fit in one file aspect of retrying).

**Comment by bersace** [comment]
@rholder :up:

**Comment by bersace** [comment]
cc @RyPeck Looks like @rholder is inactive on this project. Can you take the lead ?

**Comment by bersace** [comment]
Updated AUTHOR according to README.

**Comment by bersace** [comment]
I updated the README too.

File: retrying.py (hunk 1 above)
Comment (jpic) [reply retrying.py:48]: Could that comment be a code comment rather than a github comment ?

**Comment by jpic** [comment]
I think we'll need some documentation in the README. Otherwise, nice backflips to support multiple python versions, it's great to have them encapsulated in this lib.

Keep up the great work 

:guitar:

File: retrying.py (hunk 1 above)
Comment (bersace) [reply retrying.py:48]: You mean in the commit message ?

**Comment by bersace** [comment]
@jpic what specific documentation do we need for coroutine ? It's used exaclty the same way.

``` python
@retry
async def mycoro():
    await asyncio.sleep(1)
```

**Comment by jpic** [comment]
Maybe at least a new list item on the [features list](https://github.com/rholder/retrying#features) would be nice.

File: README.rst [hunk 3]
```diff
@@ -39,6 +39,7 @@ Features
 - Specify wait condition (i.e. exponential backoff sleeping between attempts)
 - Customize retrying on Exceptions
 - Customize retrying on expected returned result
+- Retry on coroutines
```
Comment (bersace) [inline README.rst:42]: @jpic did you saw this ?

File: retrying.py (hunk 1 above)
Comment (jpic) [reply retrying.py:48]: Yeah or just put it in a separate commit perhaps.

**Comment by bersace** [comment]
Closing in favour of https://github.com/jd/tenacity/pull/37

**Pull request closed without merging**